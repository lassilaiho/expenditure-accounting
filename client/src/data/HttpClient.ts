export class ConnectionError extends Error {
  public readonly name = 'ConnectionError';

  public get message() {
    return this.cause.message;
  }

  public constructor(public readonly cause: Error, ...params: any[]) {
    super(...params);

    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, ConnectionError);
    }
  }
}

async function apiFetch(input: RequestInfo, init?: RequestInit | undefined) {
  try {
    return await fetch(input, init);
  } catch (e) {
    if (e instanceof Error) {
      throw new ConnectionError(e);
    }
    throw e;
  }
}

export default class HttpClient {
  public baseUrl = '';

  public defaultHeaders: Record<string, string> = {
    Authorization: '',
  };

  private _sessionToken: string | null = null;
  public get sessionToken() {
    return this._sessionToken;
  }
  public set sessionToken(token: string | null) {
    this._sessionToken = token;
    this.defaultHeaders['Authorization'] =
      token === null ? '' : 'Basic ' + btoa(token);
  }

  public constructor(baseUrl: string) {
    this.baseUrl = baseUrl;
  }

  public async get(url: string) {
    return apiFetch(this.baseUrl + url, {
      mode: 'cors',
      headers: { ...this.defaultHeaders },
    });
  }

  public async postJson(url: string, body: any) {
    return apiFetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'POST',
      headers: {
        ...this.defaultHeaders,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }

  public async patchJson(url: string, body: any) {
    return apiFetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'PATCH',
      headers: {
        ...this.defaultHeaders,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }

  public async delete(url: string) {
    return apiFetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'DELETE',
      headers: { ...this.defaultHeaders },
    });
  }
}
