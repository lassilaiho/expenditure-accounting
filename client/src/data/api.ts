export type LoadState =
  | 'not-started'
  | 'loading'
  | 'finished'
  | 'failed';

export class SessionToken {
  public constructor(
    public readonly token: string,
    public readonly expiryTime: Date,
  ) { }

  public static fromJson(json: any): SessionToken | null {
    if (typeof json.token !== 'string'
      || typeof json.expiryTime !== 'string') {
      throw new Error('Invalid json object');
    }
    return new SessionToken(json.token, new Date(json.expiryTime));
  }

  public get isValid() {
    return this.expiryTime.getTime() > Date.now();
  }

  public toJSON(): unknown {
    return {
      token: this.token,
      expiryTime: this.expiryTime.toUTCString(),
    };
  }
}

export default class Api {
  public baseUrl: string = '';

  public defaultHeaders: Record<string, string> = {
    'Authorization': ''
  };

  private _sessionToken: SessionToken | null = null;
  public get sessionToken() { return this._sessionToken; }
  public set sessionToken(token: SessionToken | null) {
    this._sessionToken = token;
    this.defaultHeaders['Authorization'] =
      token === null ? '' : 'Basic ' + btoa(token.token);
  }

  public constructor(baseUrl: string) {
    this.baseUrl = baseUrl;
  }

  public async get(url: string) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      headers: { ...this.defaultHeaders },
    });
  }

  public async postJson(url: string, body: any) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'POST',
      headers: {
        ...this.defaultHeaders,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }

  public async patchJson(url: string, body: any) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'PATCH',
      headers: {
        ...this.defaultHeaders,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(body),
    });
  }

  public async delete(url: string) {
    return fetch(this.baseUrl + url, {
      mode: 'cors',
      method: 'DELETE',
      headers: { ...this.defaultHeaders },
    });
  }
}
