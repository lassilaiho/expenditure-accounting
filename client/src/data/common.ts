import { createAction } from '@reduxjs/toolkit';

import { Product } from './products';
import { Purchase } from './purchases';
import { Tag } from './tags';

export type Collection<T> = {
  byId: { [id: number]: T };
  all: number[];
};

export function emptyCollection<T>(): Collection<T> {
  return { byId: {}, all: [] };
}

export const id = <T>(x: T) => x;

export const clearRemoteData = createAction('clearRemoteData');
export type RemoteDataSet = [Purchase, Product, Tag[]][];
export const setRemoteData = createAction<RemoteDataSet>('setRemoteData');
