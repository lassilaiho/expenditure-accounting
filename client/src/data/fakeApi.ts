import Api from './api';

export type FakeResponse = (
  method: string,
  url: string,
  body?: any,
) => Promise<Response>;

export class FakeApi extends Api {
  public constructor(private fakeResponse: FakeResponse) {
    super('');
  }

  public async get(url: string) {
    return this.fakeResponse('get', url);
  }

  public async postJson(url: string, body: any) {
    return this.fakeResponse('post', url, body);
  }

  public async patchJson(url: string, body: any) {
    return this.fakeResponse('patch', url, body);
  }

  public async delete(url: string) {
    return this.fakeResponse('delete', url);
  }
}

function jsonResponse(body: any) {
  return new Response(JSON.stringify(body), {
    headers: {
      'Content-Type': 'application/json',
    },
    status: 200,
  });
}

const tags = [
  {
    id: 1,
    name: 'tag 1',
  },
  {
    id: 2,
    name: 'tag 2',
  },
  {
    id: 3,
    name: 'tag 3',
  },
];
const products = [
  {
    id: 1,
    name: 'product 1',
  },
  {
    id: 2,
    name: 'product 2',
  },
];
const purchases = [
  {
    id: 1,
    product: products[0],
    date: '2020-03-14',
    quantity: '1.0',
    price: '2.0',
    tags: [tags[0], tags[2]],
  },
  {
    id: 2,
    product: products[0],
    date: '2020-01-14',
    quantity: '10',
    price: '13.2',
    tags: [tags[2]],
  },
  {
    id: 3,
    product: products[1],
    date: '2020-03-14',
    quantity: '0.8',
    price: '0.2',
    tags: [tags[0], tags[1], tags[2]],
  },
];

export const fakeData = { tags, products, purchases };

export const fakeApi = new FakeApi(async (method, url, body) => {
  switch (method + url) {
    case 'get/purchases':
      return jsonResponse({ purchases });
  }
  return new Response(null, { status: 404 });
});
