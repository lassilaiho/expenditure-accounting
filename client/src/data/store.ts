import Big from 'big.js';
import {
  action,
  computed,
  makeObservable,
  observable,
  runInAction,
} from 'mobx';
import moment from 'moment';
import React, { useContext } from 'react';

import { ensureOk } from '../util';
import Api, { LoadState } from './api';
import * as jsonConv from './jsonConvert';

export class Tag {
  public id: number;
  public name: string;

  public constructor(id: number, name: string) {
    makeObservable(this, {
      id: observable,
      name: observable,
    });
    this.id = id;
    this.name = name;
  }

  public static fromJson(json: any) {
    jsonConv.toObject(json);
    return new Tag(jsonConv.toNumber(json.id), jsonConv.toString(json.name));
  }

  public lowerCaseMatch(s: string) {
    return this.name.toLocaleLowerCase().includes(s);
  }
}

export class Purchase {
  public id: number;
  public product: Product;
  public date: moment.Moment;
  public quantity: Big;
  public price: Big;
  public get totalPrice() {
    return this.quantity.mul(this.price);
  }
  public tags: Tag[];
  public get tagsSortedByName() {
    return this.tags.slice().sort((a, b) => {
      const aName = a.name.toLocaleLowerCase();
      const bName = b.name.toLocaleLowerCase();
      if (aName < bName) {
        return -1;
      }
      if (aName > bName) {
        return 1;
      }
      return 0;
    });
  }

  public constructor(
    id: number,
    product: Product,
    date: moment.Moment,
    quantity: Big,
    price: Big,
    tags: Tag[],
  ) {
    makeObservable(this, {
      id: observable,
      product: observable,
      date: observable.ref,
      quantity: observable.ref,
      price: observable.ref,
      totalPrice: computed,
      tags: observable,
      tagsSortedByName: computed,
    });
    this.id = id;
    this.product = product;
    this.date = date;
    this.quantity = quantity;
    this.price = price;
    this.tags = tags;
  }

  public static fromJson(json: any) {
    jsonConv.toObject(json);
    return new Purchase(
      jsonConv.toNumber(json.id),
      Product.fromJson(json.product),
      jsonConv.toMomentUtc(json.date),
      jsonConv.toBig(json.quantity),
      jsonConv.toBig(json.price),
      jsonConv.toArray(Tag.fromJson, json.tags),
    );
  }

  public lowerCaseMatch(s: string) {
    return (
      this.product.lowerCaseMatch(s) || this.tags.some(t => t.lowerCaseMatch(s))
    );
  }
}

export class Product {
  public id: number;
  public name: string;

  public constructor(id: number, name: string) {
    makeObservable(this, {
      id: observable,
      name: observable,
    });
    this.id = id;
    this.name = name;
  }

  public static fromJson(json: any) {
    jsonConv.toObject(json);
    return new Product(
      jsonConv.toNumber(json.id),
      jsonConv.toString(json.name),
    );
  }

  public lowerCaseMatch(s: string) {
    return this.name.toLocaleLowerCase().includes(s);
  }
}

export class Store {
  public dataState: LoadState = 'not-started';
  public purchases: Purchase[] = [];
  public get purchasesById() {
    const map = new Map<number, Purchase>();
    for (const p of this.purchases) {
      map.set(p.id, p);
    }
    return map;
  }
  public tagsById = new Map<number, Tag>();
  public get tags() {
    const tags: Tag[] = [];
    for (const tag of this.tagsById.values()) {
      tags.push(tag);
    }
    return tags;
  }
  private get tagsByName() {
    const map = new Map<string, Tag>();
    for (const tag of this.tagsById.values()) {
      map.set(tag.name.toLowerCase(), tag);
    }
    return map;
  }
  public productsById = new Map<number, Product>();
  public get products() {
    const products: Product[] = [];
    for (const product of this.productsById.values()) {
      products.push(product);
    }
    return products;
  }
  private get productsByName() {
    const map = new Map<string, Product>();
    for (const product of this.productsById.values()) {
      map.set(product.name.toLowerCase(), product);
    }
    return map;
  }

  public constructor(private api: Api) {
    makeObservable<
      Store,
      | 'tagsByName'
      | 'productsByName'
      | 'parsePurchaseJson'
      | 'parsePurchasesJson'
    >(this, {
      dataState: observable,
      purchases: observable,
      purchasesById: computed,
      tagsById: observable,
      tags: computed,
      tagsByName: computed,
      productsById: observable,
      products: computed,
      productsByName: computed,
      addTags: action,
      reloadData: action,
      parsePurchasesJson: action,
      parsePurchaseJson: action,
      restorePurchase: action,
    });
  }

  public async updatePurchase(id: number) {
    const purchase = this.purchasesById.get(id);
    if (purchase) {
      ensureOk(
        await this.api.patchJson(`/purchases/${purchase.id}`, {
          product: purchase.product.id,
          date: purchase.date.format(),
          price: purchase.price,
          quantity: purchase.quantity,
          tags: purchase.tags.map(t => t.id),
        }),
      );
    } else {
      throw new Error(`No purchase with id: ${id}`);
    }
  }

  public async addTags(names: string[]) {
    const newTags: string[] = [];
    const result: Tag[] = [];
    for (const name of names) {
      const tag = this.getTagByName(name);
      if (tag === undefined) {
        newTags.push(name);
      } else {
        result.push(tag);
      }
    }
    if (newTags.length > 0) {
      const resp = await this.api.postJson('/tags', { tags: newTags });
      ensureOk(resp);
      const respJson = await resp.json();
      runInAction(() => {
        for (const tagJson of respJson.tags) {
          const tag = Tag.fromJson(tagJson);
          this.tagsById.set(tag.id, tag);
          result.push(tag);
        }
      });
    }
    return result;
  }

  public async addProduct(name: string) {
    const product = this.getProductByName(name);
    if (product === undefined) {
      const resp = await this.api.postJson('/products', { name });
      ensureOk(resp);
      const product = Product.fromJson(await resp.json());
      runInAction(() => this.productsById.set(product.id, product));
      return product;
    }
    return product;
  }

  public async addPurchase(purchase: Purchase) {
    const resp = await this.api.postJson('/purchases', {
      product: purchase.product.id,
      date: purchase.date.format(),
      price: purchase.price,
      quantity: purchase.quantity,
      tags: purchase.tags.map(t => t.id),
    });
    ensureOk(resp);
    const respJson = jsonConv.toObject(await resp.json());
    const id = jsonConv.toNumber(respJson.id);
    runInAction(() => {
      purchase.id = id;
      for (let i = 0; i < this.purchases.length; i++) {
        if (purchase.date > this.purchases[i].date) {
          this.purchases.splice(i, 0, purchase);
          return;
        }
      }
      this.purchases.splice(this.purchases.length, 0, purchase);
    });
  }

  public async deletePurchase(id: number) {
    const i = this.purchases.findIndex(p => p.id === id);
    if (i < 0) {
      throw new Error(`Purchase with id ${id} does not exist`);
    }
    ensureOk(await this.api.delete(`/purchases/${id}`));
    runInAction(() => this.purchases.splice(i, 1));
  }

  public async restorePurchase(id: number) {
    const resp = await this.api.postJson(`/purchases/${id}/restore`, {});
    await ensureOk(resp);
    const purchase = this.parsePurchaseJson((await resp.json()).purchase);
    runInAction(() => {
      for (let i = 0; i < this.purchases.length; i++) {
        if (purchase.date > this.purchases[i].date) {
          this.purchases.splice(i, 0, purchase);
          return;
        }
      }
      this.purchases.splice(this.purchases.length, 0, purchase);
    });
  }

  public async getLatestPurchaseByProduct(productName: string) {
    productName = productName.toLocaleLowerCase();
    return (
      this.purchases.find(p => p.product.lowerCaseMatch(productName)) ?? null
    );
  }

  private getTagByName(name: string) {
    return this.tagsByName.get(name.toLowerCase());
  }

  private getProductByName(name: string) {
    return this.productsByName.get(name.toLowerCase());
  }

  public async reloadData() {
    try {
      this.dataState = 'loading';
      this.purchases.length = 0;
      this.tagsById.clear();
      this.productsById.clear();

      const resp = await this.api.get('/purchases');
      ensureOk(resp);

      const json = await resp.json();
      runInAction(() => {
        this.parsePurchasesJson(json);
        this.dataState = 'finished';
      });
    } catch (e) {
      runInAction(() => (this.dataState = 'failed'));
      throw e;
    }
  }

  private parsePurchasesJson(data: any) {
    const purchases = data.purchases;
    if (!Array.isArray(purchases)) {
      throw new Error('Invalid response json');
    }
    for (const obj of purchases) {
      this.purchases.push(this.parsePurchaseJson(obj));
    }
  }

  private parsePurchaseJson(data: any) {
    const purchase = Purchase.fromJson(data);
    const product = this.productsById.get(purchase.product.id);
    if (product) {
      purchase.product = product;
    } else {
      this.productsById.set(purchase.product.id, purchase.product);
    }
    for (let i = 0; i < purchase.tags.length; i++) {
      const currentTag = purchase.tags[i];
      const existingTag = this.tagsById.get(currentTag.id);
      if (existingTag) {
        purchase.tags[i] = existingTag;
      } else {
        this.tagsById.set(currentTag.id, currentTag);
      }
    }
    return purchase;
  }
}

export const StoreContext = React.createContext<Store | null>(null);

export function useStore() {
  const store = useContext(StoreContext);
  if (store === null) {
    throw new Error('Store must not be null');
  }
  if (store.dataState === 'not-started') {
    store.reloadData();
  }
  return store;
}
