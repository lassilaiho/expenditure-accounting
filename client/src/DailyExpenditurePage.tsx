import {
  AppBar,
  CircularProgress,
  Container,
  List,
  ListItem,
  ListItemText,
  Paper,
  Toolbar,
  Typography
} from '@material-ui/core';
import { History } from 'history';
import { observer } from 'mobx-react';
import moment from 'moment';
import React from 'react';
import { useHistory } from 'react-router-dom';

import { Purchase, Store, useStore } from './data/store';
import MenuButton from './MenuButton';
import { currency, reverse } from './util';

export interface DailyExpenditurePageProps {
  openNavigation: () => void;
}

const DailyExpenditurePage: React.FC<DailyExpenditurePageProps> = observer(props => {
  const store = useStore();
  const history = useHistory();
  return <>
    <AppBar position='sticky'>
      <Toolbar>
        <MenuButton onClick={props.openNavigation} />
        <Typography variant='h6'>Daily Expenditure</Typography>
      </Toolbar>
    </AppBar>
    <Container fixed>
      <Paper>
        {store.dataState === 'loading'
          ? <CircularProgress color='secondary' />
          : renderDailyExpenditure(store, history)}
      </Paper>
    </Container>
  </>;
});

function renderDailyExpenditure(store: Store, history: History<any>) {
  if (store.purchases.length === 0) {
    return <List></List>;
  }
  const dailyPurchases = store.purchases.slice();
  dailyPurchases.sort(reverse(Purchase.orderByDate));
  const result: JSX.Element[] = [];
  let prev = dailyPurchases[0];
  let expenditure = prev.totalPrice;
  for (let i = 1; i < store.purchases.length; i++) {
    const current = store.purchases[i];
    if (current.date.isSame(prev.date, 'day')) {
      expenditure += current.totalPrice;
    } else {
      result.push(makeItem(prev.date, expenditure, history));
      expenditure = current.totalPrice;
    }
    prev = current;
  }
  result.push(makeItem(prev.date, expenditure, history));
  return <List>{result}</List>;
}

function makeItem(date: moment.Moment, expenditure: number, history: History<any>) {
  const d = date.clone().local().format('D.M.YYYY');
  return (
    <ListItem
      key={date.valueOf()}
      button
      onClick={() => history.push('/expenditure/' + d)}
    >
      <ListItemText
        primary={d}
        secondary={currency.format(expenditure)} />
    </ListItem>
  );
}

export default DailyExpenditurePage;
