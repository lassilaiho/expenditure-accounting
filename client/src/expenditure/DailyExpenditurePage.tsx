import { Paper } from '@material-ui/core';
import moment from 'moment';
import React from 'react';

import { getDataState, useAppSelector, useData } from '../data/store';
import CenteredLoader from '../common/CenteredLoader';
import Scaffold from '../common/Scaffold';
import { getPurchases, Purchase } from '../data/purchases';
import GroupedExpenditureView from './GroupedExpenditureView';

const DailyExpenditurePage: React.FC = () => {
  useData();
  const dataState = useAppSelector(getDataState);
  const purchases = useAppSelector(getPurchases);
  return (
    <Scaffold
      title='Daily Expenditure'
      content={
        <Paper>
          {dataState !== 'finished' ? (
            <CenteredLoader />
          ) : (
            <GroupedExpenditureView
              purchases={purchases}
              splitHere={splitFunc}
              format={formatFunc}
            />
          )}
        </Paper>
      }
    />
  );
};
function splitFunc(prev: Purchase, current: Purchase) {
  return current.date.isSame(prev.date, 'day');
}

function formatFunc(x: moment.Moment) {
  return x.clone().local().format('D.M.YYYY');
}

export default DailyExpenditurePage;
