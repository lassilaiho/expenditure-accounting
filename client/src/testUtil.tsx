import MomentUtil from '@date-io/moment';
import { MuiPickersUtilsProvider } from '@material-ui/pickers';
import { render } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

import { fakeClient } from './data/FakeClient';
import { Api, ApiContext, AppStore } from './data/store';

export const newFakeApi = (store: AppStore) => new Api(fakeClient, store);

const TestProviders = (store: AppStore, api: Api, history?: string[]) => (
  props: any,
) => {
  return (
    <Provider store={store}>
      <ApiContext.Provider value={api}>
        <MuiPickersUtilsProvider utils={MomentUtil}>
          <MemoryRouter initialEntries={history}>{props.children}</MemoryRouter>
        </MuiPickersUtilsProvider>
      </ApiContext.Provider>
    </Provider>
  );
};

export interface CustomRenderOptions {
  store: AppStore;
  api?: Api;
  history?: string[];
}

const customRender = (ui: any, options: CustomRenderOptions) => {
  return render(ui, {
    wrapper: TestProviders(
      options.store,
      options?.api ?? newFakeApi(options.store),
      options?.history,
    ),
  });
};

export * from '@testing-library/react';

export { customRender as render };
