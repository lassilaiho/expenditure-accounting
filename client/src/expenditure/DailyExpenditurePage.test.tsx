import React from 'react';

import { newStore } from '../data/store';
import { fakeApi, render } from '../testUtil';
import DailyExpenditurePage from './DailyExpenditurePage';

const store = newStore();

test('renders correctly without data', () => {
  const { asFragment } = render(
    <DailyExpenditurePage openNavigation={() => undefined} />,
    { store },
  );
  expect(asFragment()).toMatchSnapshot();
});

test('renders correctly with data', async () => {
  await store.dispatch(fakeApi.reloadData);
  const { asFragment } = render(
    <DailyExpenditurePage openNavigation={() => undefined} />,
    { store },
  );
  expect(asFragment()).toMatchSnapshot();
});
