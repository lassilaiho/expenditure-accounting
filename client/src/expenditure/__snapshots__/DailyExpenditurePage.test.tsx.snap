// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with data 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-7"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <h6
          class="MuiTypography-root makeStyles-title-9 MuiTypography-h6"
        >
          Daily Expenditure
        </h6>
      </div>
    </header>
    <div
      class="makeStyles-contentRoot-8"
    >
      <div
        class="MuiContainer-root MuiContainer-fixed MuiContainer-maxWidthLg"
        style="width: 100%; height: 100%;"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
        >
          <ul
            class="MuiList-root MuiList-padding"
          >
            <div
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiListItemText-root MuiListItemText-multiline"
              >
                <span
                  class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                >
                  14.3.2020
                </span>
                <p
                  class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                >
                  2,16 €
                </p>
              </div>
              <span
                class="MuiTouchRipple-root"
              />
            </div>
            <div
              aria-disabled="false"
              class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiListItemText-root MuiListItemText-multiline"
              >
                <span
                  class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                >
                  14.1.2020
                </span>
                <p
                  class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                >
                  132,00 €
                </p>
              </div>
              <span
                class="MuiTouchRipple-root"
              />
            </div>
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders correctly without data 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-2"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <h6
          class="MuiTypography-root makeStyles-title-4 MuiTypography-h6"
        >
          Daily Expenditure
        </h6>
      </div>
    </header>
    <div
      class="makeStyles-contentRoot-3"
    >
      <div
        class="MuiContainer-root MuiContainer-fixed MuiContainer-maxWidthLg"
        style="width: 100%; height: 100%;"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="makeStyles-centeredContainer-5"
          >
            <div
              class="MuiCircularProgress-root MuiCircularProgress-colorSecondary MuiCircularProgress-indeterminate"
              role="progressbar"
              style="width: 40px; height: 40px;"
            >
              <svg
                class="MuiCircularProgress-svg"
                viewBox="22 22 44 44"
              >
                <circle
                  class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                  cx="44"
                  cy="44"
                  fill="none"
                  r="20.2"
                  stroke-width="3.6"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
